src/lib/content-indexing.js

```diff
@@ -0,0 +1,98 @@
+/**
+ * @fileoverview Interaction with the Content Indexing API.
+ *
+ * See https://web.dev/content-indexing-api/
+ */
+
+const CACHE_NAME = 'webdev-html-cache-v1';
+const PREFERRED_ICON_SIZE = 192;
+
+const DEFAULT_ICON = {
+  sizes: '128x128',
+  src: '/images/icon-128x128.png',
+  type: 'image/png',
+};
+
+function getIconFromImageSrc(imgSrc) {
+  if (!imgSrc) {
+    return DEFAULT_ICON;
+  }
+
+  const url = new URL(imgSrc, location.href);
+
+  // The image manipulation will only work if this icon is served by Imgix.
+  // If it's not, just use the default icon.
+  if (url.origin !== 'https://webdev.imgix.net') {
+    return DEFAULT_ICON;
+  }
+
+  url.searchParams.set('w', PREFERRED_ICON_SIZE);
+  url.searchParams.set('h', PREFERRED_ICON_SIZE);
+  url.searchParams.set('fit', 'crop');
+  url.searchParams.set('fm', 'webp');
+
+  return {
+    sizes: `${PREFERRED_ICON_SIZE}x${PREFERRED_ICON_SIZE}`,
+    src: url.href,
+    type: 'image/webp',
+  };
+}
+
+// This method syncs the currently cached media with the Content Indexing API
```

Comment:
I feel like this is ripe for a helper library.

Reply:
There is a decent amount of boilerplate code here that I've reused from other integrations. But at the same time, the specific details that differ from site to site—around URL schemes and cache storage—are pretty closely tied to the implementation. So it's been hard to abstract the reusable bits out.

If you've got any specific ideas, though, we can follow-up on that.

Reply:
FWIW, I did some internal refactoring when adding in support for on-the-fly single URL indexing.